examples/hackweek_demo/plot_pnl.py

```diff
@@ -112,93 +113,134 @@ def get_wallet_from_onchain_trade_info(
-
-    # Filter trade_data based on agent_list
-    trade_data = trade_data[trade_data["operator"].isin(agent_list)]
-
-    if len(trade_data) == 0:
-        return (pd.DataFrame([]), pd.DataFrame([]))
-
-    # %% estimate position duration and add it in
-    position_duration = max(trade_data.maturity_time - trade_data.block_timestamp)
-    # print(f"empirically observed position_duration in seconds: {position_duration}")
-    # print(f"empirically observed position_duration in minutes: {position_duration/60}")
-    # print(f"empirically observed position_duration in hours: {position_duration/60/60}")
-    # print(f"empirically observed position_duration in days: {position_duration/60/60/24}")
-    position_duration_days = round(position_duration / 60 / 60 / 24)
-    # print(f"assuming position_duration is {position_duration_days}")
-    position_duration = position_duration_days * 60 * 60 * 24
-
-    agents = trade_data["operator"].value_counts().index.tolist()
-    agent_wallets = {
-        a: Wallet(
-            address=agents.index(a),
-            balance=types.Quantity(amount=FixedPoint(0), unit=types.TokenType.BASE),
-        )
-        for a in agents
-    }
-
-    # %%
-    pnl_data: list[dict[str, float]] = []
-    time_data: list[datetime] = []  # TODO: Figure out type of timestamps. datetime?
-    for _, row in trade_data.iterrows():
-        new_pnl: dict[str, float] = {}
-        for agent in agents:
-            agent_shortname = agent[:6] + "..." + agent[-4:]
-
-            marginal_trades = pd.DataFrame(row).T
-
-            # pass in one trade at a time, and get out the updated wallet
-            agent_wallets[agent] = get_wallet_from_onchain_trade_info(
-                address=agent,
-                trades=marginal_trades,
-                index=agents.index(agent),
-                add_to_existing_wallet=agent_wallets[agent],
+Info = NamedTuple(
+    "Info", wallet=pd.DataFrame, positions=pd.DataFrame, deltas=pd.DataFrame, open_share_price=pd.DataFrame
+)
+
+
+def get_agent_info(session=None) -> dict[str, Info]:
+    """Arrange agent wallet info for plotting, while calculating open share price."""
+    if session is None:
+        session = postgres.initialize_session()
+    all_agent_wallet_info = postgres.get_wallet_info_history(session)
+    agent_info = {}
+    for agent, wallet in all_agent_wallet_info.items():
+        share_price = wallet["sharePrice"].rename({"sharePrice": "share_price"})
+
+        # Create positions
+        positions = wallet.loc[:, wallet.columns[:-2]].copy()
+        positions = positions.where(positions != 0, pd.NA)
+
+        # Create deltas
+        deltas = positions.diff()
+        deltas.iloc[0] = positions.iloc[0]
+
+        # Create NaNs of the same size as deltas
+        share_price_on_increases = pd.DataFrame(data=pd.NA, index=deltas.index, columns=deltas.columns)
+
+        # Replace positive deltas with share_price
```

Comment:
positive or negative?  looks like you're matching on negative.

Reply:
it does the opposite of what it says. it doesn't REPLACE `where`... it KEEPS `where`

so I guess you're supposed to read it as
> equals `share_price_on_increases` where `matches clause` otherwise `share_price`

I guess there's some logic to this, but I definitely read it backwards the first 50 times. it doesn't help that [np.where](https://numpy.org/doc/stable/reference/generated/numpy.where.html) although quite different is used to select and replace what matches the clause like: `share_price_on_increases.loc[np.where(deltas > 0)] = share_price`

Reply:
cool thanks